### PR TITLE
Logging in new database scheme

### PR DESCRIPTION
**Pull request by eileen-kuehn**
In MatterMiners/grafana-tardis-app#3 we defined a new database scheme we want to follow with the set of COBalD/TARDIS projects. Also the LAPIS will follow this scheme and will be the prototype implementation to test usability and completeness.

### CODE REVIEW

File: lapis/job.py [hunk 1]
```diff
@@ -117,6 +87,7 @@ def __repr__(self):
 
 
 async def job_to_queue_scheduler(job_generator, job_queue, **kwargs):
+    from lapis.utility.monitor import sampling_required
```
Comment (maxfischer2781) [inline lapis/job.py]: Is there a reason why this is imported inside the function, not at the top of the module?

File: lapis/simulator.py [hunk 2]
```diff
@@ -6,9 +6,9 @@
 
 from lapis.drone import Drone
 from lapis.job import job_to_queue_scheduler
-from lapis.utility.monitor import Monitoring, collect_pool_statistics, \
-    collect_user_demand, collect_job_statistics, collect_resource_statistics, \
-    collect_cobald_cost
+from lapis.utility.monitor import Monitoring, collect_pool_cobald_statistics, \
```
Comment (maxfischer2781) [inline lapis/simulator.py]: Is there any chance of shortening these names? Just ``pool_statistics``, or perhaps a submodule as ``cobald.pool_statistics``, for example?

File: lapis/utility/monitor.py [hunk 3]
```diff
@@ -22,94 +43,197 @@ def __init__(self, simulator: "Simulator"):
...
+                        statistic.name, record
+                    )
+
+    def register_statistic(self, statistic: Callable):
+        assert hasattr(statistic, "name") and hasattr(statistic, "logging_formatter")
+        self._statistics.append(statistic)
+
+        # prepare the logger
+        logger = logging.getLogger(statistic.name)
+        if len(logger.handlers) == 0:
```
Comment (maxfischer2781) [inline lapis/utility/monitor.py]:
```suggestion
        if not logger.handlers:
```

File: lapis/utility/monitor.py [hunk 4]
```diff
@@ -1,19 +1,40 @@
...
+    def makePickle(self, record):
+        return self.format(record).encode()
+
+
+class LoggingUDPSocketHandler(logging.handlers.DatagramHandler):
+    def makePickle(self, record):
+        return self.format(record).encode()
+
+
+class TimeFilter(logging.Filter):
```
Comment (maxfischer2781) [inline lapis/utility/monitor.py]: The class name seems misleading - it is not actually working with time. ``SimulationTimeFilter`` may be more appropriate.

You might want to drop ``Filter`` from the name completely, but it is a subjective choice between what the class does versus how it does it.

File: lapis/utility/monitor.py [hunk 5]
```diff
@@ -1,19 +1,40 @@
...
+    """
+    py:class:`TimeFilter` takes care to modify the created timestamp of a log
+    record to be set to the current simulation time.
+    """
```
Comment (maxfischer2781) [inline lapis/utility/monitor.py]: Please shorten. No need to repeat the class name, and "takes care to modify ... to be set to" seems redundant.

How about "Dummy filter that sets the log record timestamp to the simulation time"?

File: lapis/utility/monitor.py [hunk 6]
```diff
@@ -22,94 +43,197 @@ def __init__(self, simulator: "Simulator"):
...
+        tags={"tardis", "pool_configuration", "pool_type", "pool"},
+        resolution=1
+    )
+}
+
+
+def collect_pool_cobald_statistics(simulator: "Simulator") -> list:
+    """
+    Function takes care on collecting statistics of pools, i.e. allocation,
+    utilisation, demand, and supply.
```
Comment (maxfischer2781) [inline lapis/utility/monitor.py]: How about just "Collect allocation, utilisation, demand and supply of pools"?

File: lapis/utility/monitor.py [hunk 7]
```diff
@@ -22,94 +43,197 @@ def __init__(self, simulator: "Simulator"):
...
+        tags={"tardis", "parent_pool", "pool_configuration", "pool_type", "pool"},
+        resolution=1
+    )
+}
+
+
+def collect_configuration_information(simulator: "Simulator") -> list:
+    """
+    Function takes care on logging information about the configuration of
+    pools and drones, e.g. provided resources.
```
Comment (maxfischer2781) [inline lapis/utility/monitor.py]: How about "Log state changes of pools and drones"?

File: lapis_tests/utility/test_monitor.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class TestTimeFilter(object):
+    @via_usim
+    async def test_simple(self):
+        payload = {"a": "a"}
+        logger, handler = make_test_logger(__name__)
+        handler.formatter = LineProtocolFormatter(resolution=1)
+        logger.addFilter(TimeFilter())
+        async with Scope() as _:
```
Comment (maxfischer2781) [inline lapis_tests/utility/test_monitor.py]: Why are you using a Scope? There no async action performed inside.

File: lapis_tests/utility/test_monitor.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            logger.critical("message", payload)
+        _, _, _, timestamp = parse_line_protocol(handler.content)
+        handler.clear()
+        assert timestamp == 0
+        async with until(time == 10):
+            await eternity
```
Comment (maxfischer2781) [inline lapis_tests/utility/test_monitor.py]: You can also ``await (time + 10)`` instead.

File: lapis/job.py (hunk 1 above)
Comment (eileen-kuehn) [reply lapis/job.py]: Not anymore, thanks.

File: lapis/utility/monitor.py (hunk 7 above)
Comment (eileen-kuehn) [reply lapis/utility/monitor.py]: Not for this one, but the one above.

File: lapis/simulator.py (hunk 2 above)
Comment (eileen-kuehn) [reply lapis/simulator.py]: I moved the functions into different modules.

File: lapis/utility/monitor.py (hunk 4 above)
Comment (eileen-kuehn) [reply lapis/utility/monitor.py]: Renamed it to `SimulationTimeFilter`.

**Pull request merged**